aiographite/aiographitesend.py

```diff
@@ -0,0 +1,325 @@
+
+		# Sending Data
+		await self._send_message(message)
+
+
+
+	@asyncio.coroutine
+	async def send_valid_metric_dict(self, dataset: Dict, timestamp = None):
+		"""
+			Send data to graphite server when incoming data is in 'dict' format
+			@param: dataset = {
+									metric1 : value1,      // type ( string: int )
+									metric2 : value2, 
+									...
+							  }
+
+			metric1 (metric2, ...) are valid metric name for Graphite
+		"""
+		self.send_valid_metric_list(dataset.items(), timestamp)
+
+
+
+	@asyncio.coroutine
+	async def disconnect(self):
+		"""
+			Close the TCP connection 
+		"""
+		try:
+			self._writer.close()
+		except AttributeError:
+			self._writer = None
+		except Exception:
+			self._writer = None
+		finally:
+			self._writer = None
+
+
+
+	@asyncio.coroutine
+	async def close_event_loop(self):
+		"""
+			Close Event Loop. 
+			No call should be made after event loop closed
+		"""
+		self.loop.close()
+
+
+	@asyncio.coroutine
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self._reader, self._writer = yield from asyncio.open_connection(self._graphite_server, self._graphite_port, loop = self.loop)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self._graphite_server_address)
+		except Exception as e:
+			raise e
+
+
+	@asyncio.coroutine
+	async def _send_message(self, message: str) -> int:
+		"""
+			@message: data ready to sent to graphite server
+		"""
+		self._writer.write(message)
+		await self._writer.drain()
+
+
+
+	def _generate_message_for_data_list(self, dataset: List[Tuple], timestamp, formate_function, generate_message_function):
+		"""
+			generate proper formatted message 
+			@param:
+			Support two kinds of dataset
+				1)	dataset = [(metric1, value1), (metric2, value2), ...] 
+				or 
+				2)	dataset = [(metric1, value1, timestamp1), (metric2, value2, timestamp2), ...]
+		"""
+		listofData = []
+		for data in dataset:
+			# unpack metric data
+			if len(data) == 2:
+				(metric, value) = data
+			else:
+				(metric, value, data_timestamp) = data
+				timestamp = data_timestamp
+			listOfData.append(formate_function(metric, value, timestamp))
+		message =  generate_message_function(listofData)
+		return message	
+
+
+
+	def _to_graphite_valid_metric_name(self, metric_dir_list: List[str]):
+		"""
+			@purpose:
+				Make metric name valid for graphite in case that the metric name includes 
+				any special character which is not supported by Graphite
+			@example: 
+				Assuming that 
+
+					Expected_Metric_Name  =  metaccounts.authentication.password.attempted
+
+				Then input metric_dir_list should be
+
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of String
+		"""
+		return ".".join([GraphiteEncoder.encode(dir_name) for dir_name in metric_dir_list])
+
+
+
+	@property
+	def reader(self):
```

Comment:
do we want to expose the reader / writer publicly? I think we shouldn't, unless there's a good reason too.